### PR TITLE
Run copyright after running deep-copy as part of the Makefile/CI

### PR DESCRIPTION
**Pull request by dhiaayachi**
### Description

Adding a copyright header to each file is now a requirement in Consul. A serie of PRs added and updated those headers to conform the Consul BSL license.

In Consul we use [deep-copy](https://github.com/globusdigital/deep-copy) to generate deep-copy code for structures in some cases, unfortunately [deep-copy](https://github.com/globusdigital/deep-copy) do not support adding a custom header to the generated file. The lack of the header would fail the CI as part of the CI steps we verify that rerunning the deep-copy generation do not introduce any change.

To fix this, this PR add an extra step in the Makefile job for deep-copy generation to also regenerate the headers, This will have 2 effects:

- Deep copy generated files will be able to have the header and we can safely verify that nothing changed in the CI.
- Any file added without copyright header would fail the CI, as part of the verification we will regenerate the copyright headers.

### Testing & Reproduction steps

Tested locally and found few files that need to be updated.

### PR Checklist

* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

**Comment by dhiaayachi** [comment]
PS: this also include some missing copyright headers.

File: test-integ/go.mod [hunk 1]
```diff
@@ -90,7 +90,7 @@ require (
 	github.com/googleapis/gax-go/v2 v2.11.0 // indirect
 	github.com/grpc-ecosystem/go-grpc-middleware v1.3.0 // indirect
 	github.com/grpc-ecosystem/grpc-gateway/v2 v2.11.3 // indirect
-	github.com/hashicorp/consul v0.0.0-00010101000000-000000000000 // indirect
```
Comment (dhiaayachi) [inline test-integ/go.mod:93]: This is not really related but I decided to add it, as it should not matter and my IDE annoyingly update this every time I make a change and I'm pretty sure I'm not the only one, I've seen it in others setup.

File: Makefile [hunk 2]
```diff
@@ -19,6 +19,7 @@ MOG_VERSION='v0.4.0'
 PROTOC_GO_INJECT_TAG_VERSION='v1.3.0'
 PROTOC_GEN_GO_BINARY_VERSION="v0.1.0"
 DEEP_COPY_VERSION='bc3f5aa5735d8a54961580a3a24422c308c831c2'
+COPYRIGHT_TOOL_VERSION='v0.16.4'
```
Comment (analogue) [inline Makefile:22]:
```suggestion
COPYWRITE_TOOL_VERSION='v0.16.4'
```

File: build-support/scripts/devtools.sh [hunk 3]
```diff
@@ -51,6 +51,10 @@ function main {
                 codegen_install
                 return 0
                 ;;
+            -copyright )
+                copyright_install
+                return 0
+                ;;
```
Comment (analogue) [inline build-support/scripts/devtools.sh:57]:
```suggestion
            -copywrite )
                copywrite_install
                return 0
                ;;
```

File: build-support/scripts/devtools.sh [hunk 4]
```diff
@@ -149,21 +153,38 @@ function lint_install {
 }
 
 function codegen_install {
-    local deep_copy_version
-    deep_copy_version="$(make --no-print-directory print-DEEP_COPY_VERSION)"
+  deepcopy_install
+  copyright_install
```
Comment (analogue) [inline build-support/scripts/devtools.sh:157]:
```suggestion
  copywrite_install
```

File: build-support/scripts/devtools.sh [hunk 5]
```diff
@@ -149,21 +153,38 @@ function lint_install {
...
+      install_versioned_tool \
+          'deep-copy' \
+          'github.com/globusdigital/deep-copy' \
+          "${deep_copy_version}" \
+          'github.com/globusdigital/deep-copy'
+}
+
+function copyright_install {
+    local copyright_version
+    copyright_version="$(make --no-print-directory print-COPYRIGHT_TOOL_VERSION)"
```
Comment (analogue) [inline build-support/scripts/devtools.sh:173]:
```suggestion
function copywrite_install {
    local copywrite_version
    copywrite_version="$(make --no-print-directory print-COPYWRITE_TOOL_VERSION)"
```

File: build-support/scripts/devtools.sh [hunk 6]
```diff
@@ -149,21 +153,38 @@ function lint_install {
...
 
     install_versioned_tool \
-        'deep-copy' \
-        'github.com/globusdigital/deep-copy' \
-        "${deep_copy_version}" \
-        'github.com/globusdigital/deep-copy'
+        'copyright' \
+        'github.com/hashicorp/copywrite' \
+        "${copyright_version}" \
+        'github.com/hashicorp/copywrite'
```
Comment (analogue) [inline build-support/scripts/devtools.sh:179]:
```suggestion
        'copywrite' \
        'github.com/hashicorp/copywrite' \
        "${copywrite_version}" \
        'github.com/hashicorp/copywrite'
```

File: build-support/scripts/devtools.sh [hunk 7]
```diff
@@ -149,21 +153,38 @@ function lint_install {
...
 }
 
 function tools_install {
 
     lint_install
     proto_tools_install
     codegen_install
+    copyright_install
```
Comment (analogue) [inline build-support/scripts/devtools.sh:187]:
```suggestion
    copywrite_install
```

File: build-support/scripts/devtools.sh (hunk 3 above)
Comment (analogue) [inline build-support/scripts/devtools.sh:57]: Instead of having a new `--copywrite` category, can we get by with adding `copywrite` to the existing `-lint`? It looks like `deepcopy` is already under `-codegen` which makes sense.

File: Makefile [hunk 8]
```diff
@@ -433,6 +434,7 @@ deep-copy: codegen-tools ## Deep copy
 	@$(SHELL) $(CURDIR)/agent/proxycfg/deep-copy.sh
 	@$(SHELL) $(CURDIR)/agent/consul/state/deep-copy.sh
 	@$(SHELL) $(CURDIR)/agent/config/deep-copy.sh
+	copywrite headers
```
Comment (analogue) [inline Makefile:437]: Do we need to limit this to just `*.deepcopy.go` so that other files aren't caught by the check? Maybe there is cause to refactor this to check to all files unless we're already doing that somewhere else.

Comment (dhiaayachi) [reply Makefile:437]: We are not doing this to all the files right now. My thinking is that we should, as I already found out few files which don't have the copyright header.
That said, maybe it would be better to make it independent from deep-copy 🤔

File: build-support/scripts/devtools.sh (hunk 3 above)
Comment (dhiaayachi) [reply build-support/scripts/devtools.sh]: I'm not sure lint is the right category here, as lint tools are not supposed to change file content. I thought that the closest category is code-gen as we are generating file headers.

File: .github/workflows/go-tests.yml [hunk 9]
```diff
@@ -96,29 +96,28 @@ jobs:
...
-  #   - run: make --always-make deep-copy
-  #   - run: |
-  #       if ! git diff --exit-code; then
-  #         echo "Generated code was not updated correctly"
-  #         exit 1
-  #       fi
-  #   - name: Notify Slack
-  #     if: ${{ failure() }}
-  #     run: .github/scripts/notify_slack.sh
+  check-codegen:
```
Comment (analogue) [inline .github/workflows/go-tests.yml:99]: Since this is now a general "make sure copywrite headers are correct everywhere" (by using the configuration in `.copwrite.hcl`, does it make sense to rename to `check-copywrite` since it will complain about files that are not related to code generation?

Comment (dhiaayachi) [reply .github/workflows/go-tests.yml:99]: It actually does both checking that copywrite headers are inserted and deep-copy generated files are up to date.
I was thinking of separating both but it make it hard, as we will never be able to check the end state which is deep-copy file are generated and all files have the headers (including the deep-copy generated ones)

**Review by analogue — APPROVED** [review]
🚀
